cv::UMat output/input in VideoCapture/VideoWriter (data stays in GPU memory)

mikhail-nikolskiy opened the pull request:
The pull request utilizes HW acceleration added recently to FFMPEG backend of videoio module, and core APIs for interoperability with DirectX and VAAPI.
Interoperability requires OpenCL context created with DirectX/VAAPI device passed at context init time. So in case of HW decode/encode and cv::UMat, data can stay in GPU memory if VideoCapture/VideoWriter open() called before any OpenCL-related functions so OpenCL context is empty at open() time, or OpenCL context explicitly initialized in interop mode before open() via calling functions cv::directx::ocl::initializeContextFromD3D*Device/cv::va_intel::ocl::initializeContextFromVA. Otherwise data copied through system memory GPU->CPU->GPU.

<cut/>

| OpenCL context before open() | parameter PROP_HW_DEVICE | OpenCL context after open() | data stays in GPU,<br> if cv::UMat and HW mode |
| ----- | ----- | ----- | ----- |
| empty | supported | created via initializeContextFrom*, if HW mode | Yes |
| created via initializeContextFrom* | ignored, with log message if specified | unchanged | Yes
| created, but not via initializeContextFrom* | supported | unchanged | No

For example, if app creates VideoCapture then VideoWriter instances with parameters PROP_HW_ACCELERATION and PROP_HW_DEVICE before OpenCL calls, the first-opened VideoCapture will try open GPU device specified in PROP_HW_DEVICE, the second-opened VideoWriter will reuse media/OpenCL device instance previously created by VideoCapture.

The pull request also introduces new functions in ocl::Context to unify capabilities of attaching external context to OpenCL context, as it used now in three places directx.cpp (core), va_intel.cpp (core), cap_ffmpeg_hw.hpp (videoio).

-----

TODO (alalek):
- [x] move out `HAVE_MFX` of cvconfig.h
- [x] properly configure `HAVE_MFX` / `HAVE_OPENCL` for FFmpeg plugin (with include directories)
- [ ] ... `HAVE_VA` / `HAVE_VA_INTEL` / etc for FFmpeg plugin (with include directories)
- [ ] (backlog, out of scope of this PR) videoio plugins API - retrieve OpenCL buffers (currently it is limited to cv::Mat only)
- [ ] (backlog, out of scope of this PR) FFmpeg for Windows: build wrapper with enable OpenCL support and share OpenCL context
- [x] try to reuse old prebuilt videoio wrappers (like FFmpeg for Windows) in OpenCV tests (some features may missing)


### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
build_contrib:Custom=OFF
build_contrib:Custom Win=OFF

build_shared:Custom=OFF
build_examples:Custom=OFF

build_image:Linux AVX2=ubuntu:20.04

Xbuild_image:Win64=msvs2019

force_builders=Custom,Custom Win,Linux AVX2
Xbuild_image:Custom=centos:7
Xbuild_image:Custom=ubuntu:20.04
build_image:Custom=gstreamer:16.04
buildworker:Custom=linux-1,linux-4,linux-6
test_opencl:Custom=ON

build_image:Custom Win=ffmpeg
Xbuild_image:Custom Win=ffmpeg-plugin
buildworker:Custom Win=windows-2
test_modules:Custom Win=videoio
test_opencl:Custom Win=ON
```

mikhail-nikolskiy commented:
UMat output/input works in my local testing on VAAPI decoders/encoders (Ubuntu 20.04), D3D11 decoders (Windows 10), MFX decoders/encoders (Ubuntu 20.04 and Windows 10).
If GPU memory interop not possible by any reason (including below), UMat output/input works via CPU memory.
Known limitations for GPU memory interop:
* GPU memory interop not supported in videoio plugin mode
* Test system creates OpenCL context on initialization, so video acceleration tests can't trigger GPU memory interop mode unless insert context deletion 'ocl::OpenCLExecutionContext::getCurrent().release()' at start of video acceleration tests
* If MFX acceleration, GPU memory interop requires HAVE_MFX enabled. The actual dependency is only on MFX headers, no linkage dependency.
* If MFX acceleration, GPU memory interop requires custom FFMPEG build with patches for MFX+D3D11 support developed by @galinart, these patches not upstreamed to FFMPEG master yet.

mikhail-nikolskiy commented on modules/core/include/opencv2/core/ocl.hpp:
```diff
@@ -43,14 +43,16 @@
 
 CV_EXPORTS_W bool haveOpenCL();
-CV_EXPORTS_W bool useOpenCL();
+CV_EXPORTS_W bool useOpenCL(bool init = true); // TODO: can we change function signature?
```
@alalek Should be changed as we can't change signature of existent function?

mikhail-nikolskiy replied:
@asmorkalov, this PR has 4 failing tests for vp9 decoding on [Windows 10](https://build.opencv.org.cn/job/precommit/job/windows10/1294/), is it same system setup and probably same reason as [issue 19694](https://github.com/opencv/opencv/issues/19694)?

mikhail-nikolskiy commented:
@asmorkalov, this PR has 4 failing tests for vp9 decoding on [Windows 10](https://build.opencv.org.cn/job/precommit/job/windows10/1294/), is it same system setup and probably same reason as [issue 19694](https://github.com/opencv/opencv/issues/19694)?

asmorkalov commented:
@mikhail-nikolskiy Yes.

asmorkalov commented:
jenkins cn please retry a build

asenyaev commented:
jenkins cn please retry a build

asenyaev commented:
jenkins cn please retry a build

asenyaev commented:
jenkins cn please retry a build

alalek commented on modules/videoio/src/cap_ffmpeg_hw.hpp:
```diff
@@ -259,83 +274,344 @@ bool hw_check_device(AVBufferRef* ctx, AVHWDeviceType hw_type, const std::string
+    return AV_HWDEVICE_TYPE_NONE;
+}
+
+static void hw_init_opencl(AVBufferRef* ctx) {
```
There is an regression for case with disabled OpenCL subsystem in OpenCV through `OPENCV_OPENCL_RUNTIME=disabled`. Both Mat / UMat test cases don't work (skipped) with messages:

```
[OPENCV:FFMPEG:40] VAAPI driver: Intel iHD driver for Intel(R) Gen Graphics - 20.4.5 (74e2f111).
[OPENCV:FFMPEG:40] Driver not found in known nonstandard list, using standard behaviour.
[ INFO:0] global /home/alalek/projects/opencv/dev/modules/videoio/src/cap_ffmpeg_hw.hpp (268) hw_check_device FFMPEG: Using vaapi video acceleration on device: Intel iHD driver for Intel(R) Gen Graphics - 20.4.5 (74e2f111)
[ INFO:0] global /home/alalek/projects/opencv/dev/modules/videoio/src/cap_ffmpeg_hw.hpp (559) hw_create_device FFMPEG: Created video acceleration context (av_hwdevice_ctx_create) for vaapi on device 'default'
[ERROR:0] global /home/alalek/projects/opencv/dev/modules/videoio/src/cap.cpp (160) open VIDEOIO(FFMPEG): raised OpenCV exception:

OpenCV(4.5.2-dev) /home/alalek/projects/opencv/dev/modules/core/src/opencl/runtime/opencl_core.cpp:378: error: (-220:Unknown error code -220) OpenCL function is not available: [clGetPlatformIDs] in function 'opencl_check_fn'


[     SKIP ] FFMPEG VideoCapture on sample_322x242_15frames.yuv420p.libx264.mp4 not supported with HW acceleration, skipping
```

alalek commented on modules/core/include/opencv2/core/ocl.hpp:
```diff
@@ -277,6 +279,8 @@ class CV_EXPORTS Context
     void* ptr() const;
 
+    /** @returns Property specified on context creation */
+    void* getProperty(long propertyId) const;
```
As we want to work with OpenCL enums lets use more specific method name `getOpenCLContextProperty()` and add note to doc string about returned values and passed parameter (extract properties through `clGetContextInfo(CL_CONTEXT_PROPERTIES)` call)

alalek commented on modules/core/src/ocl.cpp:
```diff
@@ -3036,6 +3035,23 @@ Context Context::create(const std::string& configuration)
+    if (p == NULL)
+        return nullptr;
+    ::size_t size = 0;
+    if (CL_SUCCESS == clGetContextInfo(p->handle, CL_CONTEXT_PROPERTIES, 0, NULL, &size)) {
+        std::vector<cl_context_properties> prop(size / sizeof(cl_context_properties));
+        if (CL_SUCCESS == clGetContextInfo(p->handle, CL_CONTEXT_PROPERTIES, size, prop.data(), NULL)) {
```
Just a question: can we cache this properties array? or its contents can be changed in runtime

mikhail-nikolskiy replied:
OpenCL doesn't allow to change properties (I would really like to have such ability - change media context property after context creation;)
But properties cached by OpenCL user-mode library, so these these two calls should work very fast.
Is it ok, or better cache on OpenCV side?

mikhail-nikolskiy replied:
OpenCL doesn't allow to change properties (I would really like to have such ability - change media context property after context creation;)
But properties cached by OpenCL user-mode library, so these these two calls should work very fast.
Is it ok, or better cache on OpenCV side?

alalek replied:
Thanks for information. Caching support here is low priority for now.

alalek commented on modules/core/include/opencv2/core/ocl.hpp:
```diff
@@ -290,6 +294,21 @@ class CV_EXPORTS Context
+    };
+    template <typename T>
+    void setUserContext(std::shared_ptr<T> userContext) {
+        setUserContext(typeid(T), userContext);
```
Great idea!

> `std::shared_ptr<T>`

const reference?
inline?

alalek commented on modules/core/include/opencv2/core/ocl.hpp:
```diff
@@ -290,6 +294,21 @@ class CV_EXPORTS Context
 
+    class CV_EXPORTS UserContext {
+    public:
+        virtual ~UserContext() {};
```
It makes sense to put destructor implementation in .cpp file.
Otherwise "full header" classes may lead to RTTI issues (e.g. with clang - it creates independent clones in each module)

alalek commented on cmake/templates/cvconfig.h.in:
```diff
@@ -149,6 +149,9 @@
 #cmakedefine HAVE_VA_INTEL
 
+/* Intel MediaSDK/oneVPL */
+#cmakedefine HAVE_MFX
```
TODO(alalek): check how to move this to videoio plugins

alalek commented on modules/videoio/src/cap_ffmpeg_impl.hpp:
```diff
@@ -2210,6 +2228,45 @@ bool CvVideoWriter_FFMPEG::writeFrame( const unsigned char* data, int step, int
+
+    // check that current OpenCL context initilized on same media device as codec
+    //if (!hw_check_opencl_context(video_st->codec->hw_device_ctx))
+    //    return false;
```
Perhaps robust check should check UMat's OpenCL context.

mikhail-nikolskiy replied:
It would require code like below?
```
        CV_Assert(u.u && u.u->currAllocator == ocl::getOpenCLAllocator());
        CV_Assert(std::static_pointer_cast<ocl::Context>(u.u->allocatorContext)->ptr() == ocl_context.getContext().ptr());
```
I believe SetKernelArgument will throw exception if wrong context, and this code under try/catch - is it ok or better check context?

alalek replied:
Not sure that OpenCL CTS has checks for "wrong context" cases.

I would prefer to raise OpenCV exception instead of crash in OpenCL runtime internals.

alalek reviewed (commented):
Thank you for great contribution!

I'm sorry for huge review delay (we were busy with release activities).
This is the first review iteration.

asenyaev commented:
jenkins cn please retry a build

mikhail-nikolskiy commented:
Thanks for review!
I pushed commit addressing review comments.

Regarding "explicit VideoCapture open parameter which forces creation of new of OpenCL context".
I added such parameter preliminary named CAP_PROP_HW_DEVICE_OPENCL, and added it to tests.
Such draft naming because it could be used potentially for more functionality - force OpenCL context creation and select GPU device via value of cl_device_id type (works with all acceleration types including VIDEO_ACCELERATION_ANY), as alternative way of device selection via CAP_PROP_HW_DEVICE with index in acceleration specific enumeration.
So that user can do
```
    cl_device_id clDeviceId = ... // select device by cv::ocl calls or native OpenCL calls
    VideoCapture capture(infile, backend, {
            CAP_PROP_HW_ACCELERATION, VIDEO_ACCELERATION_ANY,
            CAP_PROP_HW_DEVICE_OPENCL, clDeviceId
    });
```
or
```
    VideoCapture capture(infile, backend, {
            CAP_PROP_HW_ACCELERATION, VIDEO_ACCELERATION_D3D11,
            CAP_PROP_HW_DEVICE, 1, // D3D11 device#1
            CAP_PROP_HW_DEVICE_OPENCL, NULL // force OpenCL context creation on device#1
    });
```
Is it beneficial?
It should be possible via device UUID comparison, but the problem that cl_device_id is pointer and properties currently don't support pointers as value.

alalek commented:
> wants to merge 155 commits

> `js: support setLogLevel() / getLogLevel() calls`
> Merge branch 'ffmpeg-umat' of https://github.com/mikhail-nikolskiy/op


Looks like "git merge" has been done in the wrong direction.
Need to merge "upstream" (external) commits, not yours.

alalek commented:
> `CAP_PROP_HW_DEVICE_OPENCL`

Looks like using some handles here is over-complicated for users.

I would prefer to have ON/OFF property:

- `CAP_PROP_HW_ACCELERATION_USE_OPENCL=OFF` (default)
    - check and re-use existed OpenCL context if context is compatible.
    - disable OpenCL buffers if context is not compatible.
- `CAP_PROP_HW_ACCELERATION_USE_OPENCL=ON`
    - check and re-use existed OpenCL context if context is compatible.
    - create new "Media" context if current context is not compatible (replaces threads's `OpenCLExecutionContext`).

mikhail-nikolskiy commented:
> Looks like using some handles here is over-complicated for users.

I renamed property to `CAP_PROP_HW_ACCELERATION_USE_OPENCL` with zero/non-zero value, and fixed merge problem.

alalek commented:
Excluded "merge" and "revert" commits from the PR (no changes in the code).
Please fetch changes from the source branch of your GitHub fork (without merge or git pull).

mikhail-nikolskiy commented:
Thanks for removing mess with revert commits!
Added fix today for DX11 problem reported by @rboriso1 - OpenCL platform with GPU+CPU devices reports only extensions supported by both GPU and CPU devices, so doesn't report DX11 interop extension although it supported by GPU. The change is to try load extension function without checking extension itself, as done already in VAAPI/DX9/DX10.

mikhail-nikolskiy commented:
On Windows 10 builder I saw test failures on test cases with use_umat=true parameter, but can't find any log messages from test execution (exceptions, errors, etc). Is it possible to get debug log? It would help a lot as can't reproduce on local iGPU and dGPU devices.

alalek commented:
> On Windows 10 builder I saw test failures on test cases with use_umat=true parameter

`Win64` / `Win64 OpenCL` builders uses prebuilt FFmpeg wrapper. It doesn't contain changes from this patch.
I need to think how to properly disable these test cases in case of old FFmpeg wrapper.

alalek reviewed (commented):
Thank you for contribution!
This patch works well in these configurations on my side (validated):
- [x] Linux + VAAPI + Intel iGPU
- [x] Windows (with FFmpeg build) + D3D11 + Intel iGPU

Remaining CI "errors" are specific to OpenCV / build configurations. I will take a look how to resolve them (added list to PR's description)

mikhail-nikolskiy commented on modules/core/src/directx.cpp:
```diff
@@ -236,187 +237,121 @@ int getTypeFromD3DFORMAT(const int iD3DFORMAT)
-        if (!is_support_cl_khr_d3d11_sharing)
-            CV_Error(cv::Error::OpenCLInitError, "OpenCL: No supported extensions");
+    ~OpenCL_D3D11_NV() {
+        device->Release();
```
Can't reproduce in my environment so far. Could be (at least) two potential reasons
1. Clean-up order - as it's after ExitProcess, is it possible that some global d3d11/dxgi resources already closed at time device Release called?
2. Incorrect manipulation (directly or indirectly) with d3d device reference counter, so that Release called on already destroyed object

I tried to check that it's not 2 (which is bad) by inserting
```
ocl::OpenCLExecutionContext::getCurrent().release()
```
into sample `video_acceleration` right before exiting function main, works ok.
For similar experiment in tests, is there easy way to close OpenCL context after all tests finished but before ExitProcess?

alalek replied:
> is there easy way to close OpenCL context after all tests finished

We can do that in our tests. But problem itself would migrate to OpenCV Users because we just "hides" that.

---

We have similar issues with OpenCL, so there are several workarounds to avoid calling of any release functions after "ExitProcess" (see `__termination` flag).

mikhail-nikolskiy replied:
I mean modify tests temporary for experimental purpose - if call to release OpenCL context before ExitProcess causes crash, it definitely requires investigation and fix; if no crash, should be ok to add workarounds targeted specifically for termination process?

alalek replied:
> for experimental purpose

Adding `ocl::OpenCLExecutionContext::getCurrent().release();` into the test's main() function eliminates crash.

---

`__termination` flag is not enabled in this problematic code flow (`DllMain()` is not called, cleanup is performed through FLS):

```
...
20 000000e6`622ff790 00007ff9`ddc6b690     opencv_core452!cv::TLSData<cv::CoreTLSData>::deleteDataInstance+0x47 [C:\build\precommit_custom_windows\opencv\modules\core\include\opencv2\core\utils\tls.hpp @ 82] 
21 000000e6`622ff7c0 00007ffa`18eeddd5     opencv_core452!cv::details::TlsStorage::releaseThread+0x130 [C:\build\precommit_custom_windows\opencv\modules\core\src\system.cpp @ 1571] 
22 000000e6`622ff800 00007ffa`18eedae8     ntdll!RtlpFlsDataCleanup+0x121
23 000000e6`622ff840 00007ffa`18eeda2d     ntdll!LdrShutdownProcess+0x98
24 000000e6`622ff940 00007ffa`1731e0ab     ntdll!RtlExitUserProcess+0xad
25 000000e6`622ff970 00007ffa`16a405bc     KERNEL32!ExitProcessImplementation+0xb
26 000000e6`622ff9a0 00007ffa`16a4045f     ucrtbase!exit_or_terminate_process+0x44
27 000000e6`622ff9d0 00007ff7`81b0718b     ucrtbase!common_exit+0x6f
28 000000e6`622ffa20 00007ffa`17317034     opencv_test_videoio!__scrt_common_main_seh+0x173 [D:\agent\_work\9\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 295] 
29 000000e6`622ffa60 00007ffa`18ee2651     KERNEL32!BaseThreadInitThunk+0x14
2a 000000e6`622ffa90 00000000`00000000     ntdll!RtlUserThreadStart+0x21
```

Will take a look how to deal this that case.

dmatveev commented:
That's interesting stuff for sure!

@mikhail-nikolskiy would you mind adding support for `cv::RMat` & `cv::MediaFrame` as well? Those data structures are especially designed to handle the remote memory with custom adaptors inside (e.g. this way we to zero-copy for GStreamer buffers in the G-API pipeline), so may be worth adding too. The latter `cv::MediaFrame` structure brings more structured handling for the media formats like NV12 (and other YUV variants can be added there, too).

mikhail-nikolskiy commented:
> @mikhail-nikolskiy would you mind adding support for `cv::RMat` & `cv::MediaFrame` as well?
I checked RMat::access, it seems to be functionally equal to UMat::getMat, so wrapper from UMat to RMat could be implemented in generic way, not specifically in videoio module?
(Dmitry, your question actually triggers important topic around MatAllocator/Adaptor concepts, but probably better discuss separately)

alalek commented:
@mshabunin Could you please take a look on CMake scripts changes?

**Note**: "plugins" build mode is not fully supported yet.

mshabunin commented on CMakeLists.txt:
```diff
@@ -1420,7 +1420,16 @@ if(WITH_LIBREALSENSE OR HAVE_LIBREALSENSE)
-  status("    Intel Media SDK:" HAVE_MFX      THEN "YES (${MFX_LIBRARY})" ELSE NO)
+  if(HAVE_MFX)
+    if(MFX_LIBRARY)
+      set(__details " (${MFX_LIBRARY)")
```
```suggestion
      set(__details " (${MFX_LIBRARY})")
```
A curly brace is missing here.

mikhail-nikolskiy commented on modules/videoio/src/cap_ffmpeg_hw.hpp:
```diff
@@ -80,7 +113,7 @@ std::string getDecoderConfiguration(VideoAccelerationType va_type, AVDictionary
     case VIDEO_ACCELERATION_D3D11: return "";
     case VIDEO_ACCELERATION_VAAPI: return "vaapi.iHD";
-    case VIDEO_ACCELERATION_MFX: return "";
+    case VIDEO_ACCELERATION_MFX: return "qsv.iHD";
```
No easy workaround. ffmpeg-qsv decoder designed with delayed initialization, the `MFXVideoDECODE_Init` called on decoding first frame, it's too late for fallback to SW decoder in current implementation. No such problem in VAAPI and D3D11 decoders as they report all errors on opening stage (av_codec_open2 call).
Disable MFX decode (revert "qsv.iHD" back to empty string "") or keep as known issue in case MFX acceleration explicitly requested?

alalek replied:
@mikhail-nikolskiy Thank you for the information!

Looks like we need to add the first frame grabbing as part of `.open()` implementation to handle/workaround this case (I believe we can implement this later).

alalek reviewed (approved):
Great contribution! Thank you :+1:

mshabunin commented on modules/videoio/cmake/detect_ffmpeg.cmake:
```diff
@@ -99,6 +99,38 @@ if(HAVE_FFMPEG_WRAPPER)
+    elseif(OPENCL_INCLUDE_DIR)
+      set(__opencl_dirs "${OPENCL_INCLUDE_DIR}")
+    else()
+      set(__opencl_dirs "${OpenCV_SOURCE_DIR}/3rdparty/include/opencl/1.2")
```
I observe cmake error in some configurations:
```
CMake Error in modules/CMakeLists.txt:
  Target "ocv.3rdparty.ffmpeg.builtin_deps" INTERFACE_INCLUDE_DIRECTORIES
  property contains path:

    "/opencv/3rdparty/include/opencl/1.2"

  which is prefixed in the source directory.
```
It seem to be related to installation process: https://stackoverflow.com/questions/25676277/cmake-target-include-directories-prints-an-error-when-i-try-to-add-the-source

alalek replied:
Thanks, this problem is caused by `-DBUILD_SHARED_LIBS=OFF` option.
